src/mkFlake/fup-adapter.nix

```
@@ -51,20 +51,71 @@ let
 
   # evalArgs sets channelName and system to null by default
   # but for proper default handling in fup, null args have to be removed
-  stripHost = args: removeAttrs (lib.filterAttrs (_: arg: arg != null) args) [
+  stripNull = args: (lib.filterAttrs (_: arg: arg != null) args);
+
+  stripHost = args: removeAttrs (stripNull args) [
     # arguments in our hosts/hostDefaults api that shouldn't be passed to fup
     "externalModules" # TODO: remove deprecated option
     "exportedModules"
     "tests"
   ];
 
+  nixosHostDefaults = flake-utils-plus.lib.mergeAny
+    {
+      system = "x86_64-linux";
+      output = "nixosConfigurations";
+
+      # add `self` & `inputs` as specialArgs so their libs can be used in imports
+      specialArgs = config.nixos.importables // { inherit (config) self inputs; };
+
+      modules = config.nixos.hostDefaults.exportedModules ++ defaultHostModules ++ [
+        ({ config, pkgs, self, ... }: {
+          users.mutableUsers = lib.mkDefault false;
+          hardware.enableRedistributableFirmware = lib.mkDefault true;
+          system.configurationRevision = lib.mkIf (self ? rev) self.rev;
+        })
+      ];
+    }
+    (stripNull config.nixos.hostDefaults);
+  nixosHosts = lib.mapAttrs
+    (
+      _: hostConfig:
+        flake-utils-plus.lib.mergeAny
+          nixosHostDefaults
+          (stripNull hostConfig)
+    )
+    config.nixos.hosts;
+
+  darwinHostDefaults = flake-utils-plus.lib.mergeAny
+    {
+      system = "x86_64-darwin";
+      output = "darwinConfigurations";
+      builder = darwin.lib.darwinSystem;
+
+      # add `self` & `inputs` as specialArgs so their libs can be used in imports
+      specialArgs = config.darwin.importables // { inherit (config) self inputs; };
+      modules = config.darwin.hostDefaults.exportedModules ++ defaultHostModules;
+    }
+    (stripNull config.darwin.hostDefaults);
+  darwinHosts = lib.mapAttrs
+    (
+      _: hostConfig: flake-utils-plus.lib.mergeAny
+        darwinHostDefaults
+        (stripNull hostConfig)
+    )
+    config.darwin.hosts;
+
   diggaFupArgs = {
     inherit (config)
       channelsConfig
       supportedSystems;
     inherit self inputs sharedOverlays;
 
-    hosts = builtins.mapAttrs (_: stripHost) config.nixos.hosts;
+    hosts = builtins.mapAttrs (_: stripHost) (
+      flake-utils-plus.lib.mergeAny
+        nixosHosts
+        darwinHosts
```

```suggestion
      flake-utils-plus.lib.mergeAny
        nixosHosts
        darwinHosts
```
While I realize that this makes the code easier, the issue with this is that darwin and nixos hosts can't have the same name even though this is actually possible. This creates an artificial restriction on host names. If we do decide to do this, there needs to be an error triggered early when a user passes a host to both `nixos.hosts` and `darwin.hosts` so they know that only one will be created.

---

Good point, though I assumed that would be an edge case. There is a unique-name restriction amongst `nixos.hosts` already, and I would think that allowing identical hostnames across `{nixos,darwin}.hosts` would be difficult to manage and could lead to weird issues down the road. So I'm thinking that throwing an error for any identical hostnames would be the better option here.